crates/ra_hir_ty/src/traits/chalk.rs

```diff
@@ -448,11 +490,22 @@ pub(crate) fn fn_def_datum_query(
     let where_clauses = convert_where_clauses(db, callable_def.into(), &bound_vars);
     let bound = rust_ir::FnDefDatumBound {
         // Note: Chalk doesn't actually use this information yet as far as I am aware, but we provide it anyway
-        argument_types: sig.value.params().iter().map(|ty| ty.clone().to_chalk(db)).collect(),
-        return_type: sig.value.ret().clone().to_chalk(db),
+        inputs_and_output: make_binders(
+            rust_ir::FnDefInputsAndOutputDatum {
+                argument_types: sig
+                    .value
+                    .params()
+                    .iter()
+                    .map(|ty| ty.clone().to_chalk(db))
+                    .collect(),
+                return_type: sig.value.ret().clone().to_chalk(db),
+            },
+            sig.num_binders,
```

Comment:
creating two levels of binders in chalk for the same binders on r-a side seems highly suspicious to me. TBH I'm not completely sure what these inner binders are for (maybe something lifetime-related, in which case it would be fine to use 0 binders here. We need to shift in all the types here another level though.

Reply:
Ah yeah, these inner binders are for late-bound regions. So it should be fine to always set them to 0 (but we still need to shift in, one `shifted_in()` on the whole `FnDefInputsAndOutputDatum` structure should work)